R/arima.R

```diff
@@ -330,49 +377,56 @@ do_arima <- function(df, time,
     # Reference: https://github.com/tidyverts/fable/issues/91
     ret <- tibble(data = list(ret_df), model = list(model_df))
 
+    tryCatch({
+      stl_ts <-ts(training_data$y, frequency=seasonal_periods)
+      stl_res <- stl(stl_ts, "periodic")
+      stl_df <- as.data.frame(stl_res$time.series)
+      stl_df[[time_col]] <- training_data$ds
+
+      # Detect change points
+      cpt_res <- changepoint::cpt.var(training_data$y, method="PELT")
+      cpt_vec <- rep(0, length(training_data$ds))
+      cpt_vec[cpt_res@cpts] <- 1
+      cpt_vec[length(cpt_vec)] <- 0 # Last data point sometimes is reported as a change point, but this is not useful for our purpose.
+      stl_df$change_point <- cpt_vec
+      stl_df$y <- training_data$y
+
+      stl_seasonal_df <- stl_df %>% dplyr::slice(1:seasonal_periods) # To display only one seasonal cycle
+      ret <- ret %>% mutate(stl = list(!!stl_df), stl_seasonal = list(!!stl_seasonal_df))
+    }, error = function(e) { # This can fail depending on the data.
+      # At least, create stl, stl_seasonal columns to avoid error.
+      ret <<- ret %>% mutate(stl = list(data.frame()), stl_seasonal = list(data.frame()))
```

Comment:
Is `ret` global variable?

Reply:
No, but <<- is needed since it is inside function(e){...}.

Reply:
I got it.